### DISCUSSION TITLE
Define domain with multiple material functions

### DISCUSSION THREAD

**Discussion by TLWise** [idaholab/moose#23694] [Q&A Modules: Navier-Stokes] [answered]
I am currently using, type = ADGenericConstantMaterial,   prop_names = 'rho mu cp k'. I would like to define each of the materials with an analytical function. is there something similar to ADGenericConstantMaterial that allows me to define rho mu cp k, with an analytical function for each?

**Comment by GiudGiud** [comment, marked as answer]
Hello

you can use GenericFunctionMaterial for functions of time and space. 

For analytical functions of things like variables you may use the ParsedMaterial

Guillaume

**Reply by TLWise** [reply]
Thank you very much. I thought that GenericFunctionMaterial might be the used for that, but I was not sure of the syntax and the ability to define a function for multiple entities. Thanks again!

**Reply by GiudGiud** [reply]
Yeah it can. The big limitation to functions is that, without additional code, are only functions of space and time

**Reply by TLWise** [reply]
What do most users use when they want to vary material properties as functions of defined variables?

**Reply by GiudGiud** [reply]
If you have a simple expression you can write in an input files, the ParsedMaterial

if not then you ll have to make your own material, using C++ code. The tutorial explains how

**Reply by TLWise** [reply]
It appears that these (Material property 'rho', requested by 'ins_mat' is not defined on block Coolant
Material property 'rho', requested by 'ins_mat_face' is not defined on block Coolant
Material property 'rho', requested by 'ins_mat_neighbor' is not defined on block Coolant
Material property 'rho', requested by 'mass_pspg' is not defined on block Coolant
) expect that material properties be defined in from ADGenericConstantMaterial is preventing my use of ParsedMaterial to specify rho as a function of temperature.

**Reply by GiudGiud** [reply]
You can only specify a property once. 

You also need to be consistent with using AD. So you ll want to use an ADParsedMaterial most likely

**Reply by TLWise** [reply]
Yep. I figured that would probably be the cause for the issue. I tried the command using AD, and it is what resulted in the previous issue I shared. I removed rho as a definition in the ADConst defined it in the ADParsed. The other portions of AD did not recognize rho as being defined.

**Reply by GiudGiud** [reply]
can you please paste your former and current Materials block?

**Comment by TLWise** [comment]
```
[Materials] # Old
  [./const]
    type = ADGenericConstantMaterial
    prop_names = 'rho mu cp k'
    prop_values = '1000  0.001 4179.0  0.598' #[kg/m^3*1e-9] [Pa*sec*1e-6] [J/kg/K*1e6] [W/m/K*1e3] non exponential forms
  [../]
  [ins_mat]
    type = INSADStabilized3Eqn
    velocity = velocity
    pressure = p
    temperature = T_fluid
  []
[]
```

```
[Materials] # New
  [./const]
    type = ADGenericConstantMaterial
    prop_names = 'mu cp k'
#    prop_values = '0.001 4179.0  0.598' #[kg/m^3*1e-9] [Pa*sec*1e-6] [J/kg/K*1e6] [W/m/K*1e3] non exponential forms 
  [../]
  [./rho_density]
    type = ADParsedMaterial
    property_name = 'rho'
    coupled_variables = 'T_fluid'
    function = '99.98*((1.000088*(T_fluid-283.2))/(1-(p-1.0e5)/2.15e9))'
#    outputs = exodus
  [../]
  [ins_mat]
    type = INSADStabilized3Eqn
    velocity = velocity
    pressure = p
    temperature = T_fluid
  []
[]
```

**Reply by GiudGiud** [reply]
what error are you getting?

This looks right, except it should be: `expression = '99.98*((1.000088*(T_fluid-283.2))/(1-(p-1.0e5)/2.15e9))'`

**Reply by GiudGiud** [reply]
oh is your MOOSE old?

if so `property_name` is not the name of the parameter, it s still called f_name in an old version of MOOSE
property_name would return an unused parameter warning

**Reply by TLWise** [reply]
Yes, the version on this system is old, and I have to switch between expression and function when moving from one computer to the other.

**Reply by GiudGiud** [reply]
You ll have to use the old parameters. property-name is not recognized, so the default for f_name is being used.
so instead of declaring `rho` you are declaring a property as `F` and naturally getting an error

**Reply by TLWise** [reply]
This is the error that I get when I use T_fluid in the definition of rho:

*** ERROR ***
Invalid function
99.98*((1.000088*(T_fluid-283.2))/(1-(p-1.0e5)/2.15e9))

in ParsedMaterialHelper.
Syntax error: Unknown identifier

**Reply by GiudGiud** [reply]
you need to adapt all the parameters
All your neew syntax parameters are not recognized

you should have `args = 'p T_fluid'`

**Comment by TLWise** [comment]
My apologies upfront for not having mastered how to prevent the super large fonts.

**Reply by GiudGiud** [reply]
no worries. I use the triple backquotes to do that

**Comment by TLWise** [comment]
Had to remove the reference to T_fluid, and it got me further in the processing chain. I now get this error instead:

*** ERROR ***
/home/twise/projects/moose/modules/combined/7x7PWR_Assembly/Models/CoolantVelocityPWR.i:217: unused parameter 'Materials/rho_density/coupled_variables'

**Reply by GiudGiud** [reply]
should be args, not coupled_variables. 
You can add T and p back once you have that fixed
